### PR TITLE
Support uploading DARs to the trigger service

### PR DESCRIPTION
**Pull request by cocreature**
This PR adds a new `upload_dar` endpoint that accepts a DAR as a
multi-part form request and adds it to the list of compiled packages.

I’ve also made the DAR passed in on startup optional now given the new
endpoint.

There is no endpoint for deleting a DAR so far but there is none on
the ledger API either so I think this not particularly urgent.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala [hunk 1]
```diff
@@ -212,30 +202,83 @@ object Server {
...
+                    complete(uuid.toString)
 
-              }
+                }
+            }
+          },
+          // upload a DAR as a multi-part form request with a single field called
+          // "dar".
+          path("upload_dar") {
+//            extractExecutionContext { implicit ec =>
```
Comment (aherrmann-da) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala]: Is this still needed?

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala [hunk 2]
```diff
@@ -212,30 +202,83 @@ object Server {
...
+                          }
+                          addDar(compiledPackages, dar)
+                          complete(s"DAR uploaded, main package id: ${dar.main._1}")
+                        } catch {
+                          case e: ParseError => complete((StatusCodes.UnprocessableEntity, e))
+                        }
+                    }
+                }
+              // foreach { byteString =>
+              //   println("GOT REQUEST")
```
Comment (aherrmann-da) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala]: Debug left-over?

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala [hunk 3]
```diff
@@ -212,30 +202,83 @@ object Server {
...
+            }
           }
-        }
+          //        }
```
Comment (aherrmann-da) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala]: Same here

File: triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/ServiceTest.scala [hunk 4]
```diff
@@ -95,40 +99,74 @@ class ServiceTest extends AsyncFlatSpec with Eventually {
...
+      _ <- assert(resp.status == StatusCodes.UnprocessableEntity)
+      resp <- uploadDar(uri, darPath)
+      body <- resp.entity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String)
+      _ <- body should startWith("DAR uploaded")
+      resp <- startTrigger(uri, s"${dar.main._1}:TestTrigger:trigger", "Alice")
+      _ <- assert(resp.status.isSuccess)
+      triggerId <- resp.entity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String)
+      resp <- stopTrigger(uri, triggerId)
+      _ <- assert(resp.status.isSuccess)
+      // upload dar
```
Comment (aherrmann-da) [inline triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/ServiceTest.scala]: That comment seems misplaced

File: triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/ServiceTest.scala [hunk 5]
```diff
@@ -95,40 +99,74 @@ class ServiceTest extends AsyncFlatSpec with Eventually {
...
+        resp <- startTrigger(uri, s"${dar.main._1}:TestTrigger:trigger", "Alice")
+        triggerId <- {
+          assert(resp.status.isSuccess)
+          resp.entity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String)
+        }
+        // Trigger is running, create an A contract
+        _ <- {
+          val cmd = Command().withCreate(
+            CreateCommand(
+              templateId = Some(Identifier(dar.main._1, "TestTrigger", "A")), // template id
```
Comment (aherrmann-da) [inline triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/ServiceTest.scala]:
```suggestion
              templateId = Some(Identifier(dar.main._1, "TestTrigger", "A")),
```
seems redundant.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala (hunk 2 above)
Comment (cocreature) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala]: Good catch, I clearly forgot the cleanup pass for this PR. Sorry :slightly_smiling_face: Should all be fixed now.

Comment (aherrmann-da) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala]: No worries, thanks for fixing these!

**Pull request merged**